Draft PR: Ecash (#1093)
BitcoinDesign/Guide: guide/how-it-works/ecash/design-best-practices.md

It's important to inform users that auto-swapping involves fees and that the amount depends on the amount of ecash they want to swap and the current network conditions.

{% include tip/close.html %}

### Backup and restore for cashu
The Cashu backup and restoration process is designed to ensure users can securely recover their wallets and maintain access to their bitcoin backed ecash tokens in the event they switch devices or experience data loss. **Users should only use their recovery seed phrase once.** Repeated use of the seed phrase for restoration can lead to synchronization issues and potential errors. This is due to the fact that each time you restore, you might be dealing with an outdated state of the wallet, which can cause discrepancies in token balances and transactions.

{% include tip/open.html color="blue" icon="info" label="Single Use Recovery" %}

After successfully restoring your wallet using a recovery seed phrase, it is highly recommended that you generate a new recovery seed phrase immediately. This step is crucial because the original recovery phrase has now been used and could be more susceptible to synchronization risks.

> 

Reviewer: Sync issues would happen only if the seed is being used in two or more places. If its not being used in the old place anymore it would be okay. Though if the seed is potentially compromised it should not be used anymore.

Author: So, if I understand correctly, reusing the seed phrase more than once is fine as long as there isn’t another active instance of the wallet elsewhere. For example:

A sync issue **won’t** occur when:
1. The user factory resets their wallet and restores it using the seed phrase.
3. Two weeks later, the user factory resets their wallet again.
4. The user restores their wallet using the same seed phrase.

Reviewer: right

Author: As long as they don't have two instances of their wallet then using the same seed phrase should be okay. I'll update this when I can.

{% include tip/close.html %}

### Backup process
{% include image-gallery.html pages = page.images_backup %}

> 1. **Deterministic Wallet with Seed Phrase**: Cashu uses a deterministic wallet model, where all cryptographic keys and bitcoin backed ecashtokens can be derived from a single seed phrase. This seed phrase can be generated when the wallet is first created, or a at any point in time when the user wants to backup their wallet.

Reviewer: The seed can't be generated after the fact as previously generated could not be derived from it. In this case the wallet would have to swap proofs with secrets generated in some other way from proofs with secrets derived from the seed.

Author: When you say “after the fact,” do you mean after a user has used the wallet to get ecash? Does this mean the user _must_ generate and back up their seed phrase when the Cashu wallet is created, unlike Bitcoin wallets where you can back up your wallet at any time?

Reviewer: Only ecash derived from the seed can be restored later from that seed. So it depends what you mean by back up. A wallet could create a seed when the wallet is created but not show the user until they decide to go back it up later but be using the seed in the background to derive the secrets for the ecash. In this case the user can back the seed up later the still restore as the seed was being used from the start they just didn't know. 

But if a wallet is created without a seed and secrets are derived randomly and then later the user creates a seed and backs it up the randomly derived secrets could not be restored from the seed.  This is also true for bitcoin wallets. 

I think the best ux is for the wallet to generate a seed and prompt the user back it up when the wallet it created but **not** enforce them to back it by reentering the words. And then maybe at a set period or if the wallet balance reaches a certain amount prompt again to backup until the users confirm they did.

Author: Agreed. The practice of having the wallet generate a backup at the time of wallet creation is the pattern we should follow. I wasn’t aware this applied to bitcoin wallets as well, where if a seed is created after users start using the wallet, any funds received prior to seed creation would be lost. I'm guessing most bitcoin wallets communicating al this to the user by automatically generating a seed phrase during wallet creation. It’s up to the wallet’s UX to decide whether to present the seed phrase backup process during onboarding or let users handle it later.

2. **Secure Storage**: Users are advised to store their seed phrase securely. It is crucial to keep this seed phrase in a safe, physical format (such as written on paper) and stored in a secure location to prevent unauthorized access or loss. The best practices in the [Backup & recovery]({{ '/guide/daily-spending-wallet/backup-and-recovery/landing-page/' | relative_url }}) section also apply to Cashu wallets. 

{% include tip/open.html color="blue" icon="info" label="Mint Information Needed" %}

> Keep a record of the mints you are connected to. During the restoration process, you will need to reconnect to these specific mints to re-verify your tokens. Without this information, you might not be able to fully restore your wallet or verify your tokens.
> 

Reviewer: ```suggestion
Keep a record of the mints you are connected to. During the restoration process, you will need to reconnect to these specific mints to re-verify your tokens. Without this information, you will not be able to fully restore your wallet.

```

{% include tip/close.html %}

### Restoration process
{% include image-gallery.html pages = page.images_restore %}

> 1. **Seed Phrase Entry**: When restoring a wallet, users enter their seed phrase into their wallet. This seed phrase regenerates all cryptographic keys associated with the user's account.

Reviewer: ```suggestion
1. **Seed Phrase Entry**: When restoring a wallet, users enter their seed phrase into their wallet. This seed phrase regenerates secrets that may have been used to mint ecash.
```

> 2. **Mint(s) Entry**: Users must input the mint URL for each mint they used before. This ensures their wallet can properly reconnect and verify their tokens.

Reviewer: ```suggestion
2. **Mint(s) Entry**: Users must input the mint URL for each mint they used before. This ensures their wallet can properly reconnect and recreate their tokens.
```

> 3. **Verification**: The mint(s) checks these proofs against its records to confirm that the bitcoin backed ecash tokens have not been previously spent and are still valid. This process relies on the signatures (proofs) that were initially generated when the tokens were minted.

Reviewer: ```suggestion
3. **Verification**: The mint(s) checks these blinded secrets against its records to confirm if it has been seen before and can reissue the signature.
```

{% include tip/close.html %}

### Cashu metadata fields: explanation and usage

[NUT06](https://github.com/cashubtc/nuts/blob/main/06.md) provides a standardized way to display metadata for Cashu mints. Incorporating these metadata fields makes your application more transparent and informative. Here's a breakdown of each field and how it can be used in wallet design:

#### Mint Identity and Version

<div class="center" markdown="1">

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/cashu-id1.png"
   retina = "/assets/images/guide/how-it-works/ecash/cashu-id1@2x.png"
   modalImage = "/assets/images/guide/how-it-works/ecash/cashu-id1@2x.png"
   alt-text = "Mint identity and version display in a wallet interface"
   width = 250
   height = 400
   layout = "float-right-desktop -background -shadow"
   caption = "Consider the most important information to display to the user about the mint."
%}

- `name`: The name of the mint. Display this prominently in mint listings and details pages to help users quickly identify different mints.

> - `pubkey`: The hex public key of the mint. While this may not be directly displayed to users, it's crucial for security verification. Consider showing a shortened version or QR code for advanced users.

Reviewer: The pubkey actually isn't used for anything in production at the moment

Author: Noted! Removed from the guide for now.

- Use the `nuts` information to create a features list that helps users understand each mint's capabilities.

By effectively incorporating these Cashu metadata fields, you can create a more transparent and informative experience for users, helping them make informed decisions about which Cashu mints to trust and use. This is particularly important for Cashu, as users interact directly with individual mints rather than a federation structure like in Fedimint.

### Edit mint URL & refresh mint settings

{% include image-gallery.html pages = page.images_mint-url-edit %}

It's useful to provide users with the ability to manually update or refresh mint settings in their wallets. While best practices suggest that wallets should auto-refresh and update mint settings periodically, there may be cases where this isn't supported, or a mint undergoes significant updates, such as a URL change. By allowing users to edit the mint URL or refresh settings manually, you ensure that users always have the latest mint configurations.

### Pay to public key (P2PK)

{% include image-gallery.html pages = page.images_p2pk %}

> [NUT11](https://github.com/cashubtc/nuts/blob/main/11.md) is a powerful feature that allows bitcoin backed ecash tokens to be securely locked to another user's public key, which is generated by the recipient's wallet. This ensures that only the intended recipient can redeem the ecash. NUT11 enables secure offline payments, preventing double-spending. Beyond these basics, NUT11 supports advanced use cases like timelocks and multisignature (multisig) setups, where ecash can be conditionally spent or jointly owned by multiple parties. When designing make sure these functionalities are clearly communicated to users, highlighting their practical benefits and flexibility.

Reviewer: It might be worth noting that P2PK ecash is not derived from a seed phrase and thus cant be backed up via the seed phrase.

Author: Do you mean the actual ecash tokens that are locked with P2PK or the public key that a user generates?

Reviewer: The actual proofs of P2PK or any nut10 secret are not derived from the secret so they are not backed up by the secret.

Author: This is important! I added a tip underneath this section: P2PK ecash tokens are not derived from a seed phrase. This means they cannot be backed up using a seed phrase, and in the event of a recovery from seed phrase the P2PK ecash tokens will be lost.

---
layout: guide
title: Design best practices
description: An overview on best design practices for bitcoin-backed ecash.
nav_order: 5
parent: Ecash
grand_parent: How it works
permalink: /guide/how-it-works/ecash/design-best-practices
redirect_from:
 - /guide/how-it-works/ecash/
main_classes: -no-top-padding
image: https://bitcoin.design/assets/images/guide/how-it-works/ecash/best-practicies.jpg
image_base: /assets/images/guide/how-it-works/ecash/
images_mint-list:
    - file: federation-list
      modalImage: federation-list@2x
      alt: List of federations with QR codes for easy scanning and joining.
      caption: Federations with QR codes. Each QR code contains the federation's URL, allowing users to display, and invite another person to scan and join.
    - file: cashu-mint-list
      modalImage: cashu-mint-list@2x
      alt: List of Cashu mints with mint addresses displayed.
      caption: Cashu mints with mint addresses displayed.

images_pending:
    - file: pending-token
      modalImage: pending-token@2x
      alt: A transaction history that shows received, pending, and sent transaction each with their own visual indicator.

>       caption: Received, pending, and sent transaction each with their own visual indicator.

Author: Wallets can see the status of tokens that have been generated/issued. The token will show as 'pending' until it is redeemed by someone.

    - file: pending-transaction-details
      modalImage: pending-transaction-details@2x

>       alt: Expanded view of a pending transaction, showing the memo, amount, and the mint that created the token.

Author: Yes, it's possible for a token to be received but not yet redeemed if the recipient is offline. I'll need to think of some alternative text to indicate that. Nothing immediately comes to mind other than the term 'pending.'

>       caption: Expanded view of a pending transaction. While token is pending, anyone with the token string can claim it.

Author: Yeah this is super important. I moved it from the caption and into the paragraph below.

    - file: pending-transaction-qr
      modalImage: pending-transaction-qr@2x
      alt: A QR code for a pending transaction with the token string available to copy to clipboard.

>       caption: QR code for a pending transaction.

Author: Token is correct, thanks for the catch.

images_mint-default:
    - file: cashu-mint-details
      modalImage: cashu-mint-details@2x

>       alt: Cashu Mint metadata fields displayed in a wallet interface.

Author: IIRC, it’s not possible to know that information at the mint level. ‘Last connected’ isn’t very helpful, in my opinion, because it only shows the last time your wallet checked the mint’s status. The mint could have been offline for months, but if your wallet checked yesterday and noticed it was offline, that wouldn’t provide much insight. You can only determine whether it’s currently online or offline. With Fedimint, you can see when a Guardian was last online, but with Cashu, you can’t.

You’re right that we should show what we do know, so I added status indicators to display whether a mint is online or offline.

      caption: Cashu Mint metadata fields displayed in a wallet interface.

>     - file: cashu-mint-details-settings

Author: Updated

      modalImage: cashu-mint-details-settings@2x
      alt: A mobile ecash wallet with a bottom sheet that lists the following actions, refresh mint settings, update mint url, set as default mint, and remove mint.
      caption: Advance mint operations like setting a default mint, updating mint url, and removing a mint.

>     - file: cashu-mint-details-default-set-confirmation

Author: updated the bottom sheet copy so it explains in more detail what this action does:
_"This mint will now be used for auto-swapping received tokens and as the preferred mint for sending payments."_

      modalImage: cashu-mint-details-default-set-confirmation@2x
      alt: A success confirmation that the mint has been set as default.
      caption: Confirmation when a mint has been set as default.
    - file: cashu-default-applied
      modalImage: cashu-default-applied@2x
      alt: A list of mints with the default mint highlighted by a home icon.

>       caption: Default mint highlighted by a home icon.

Author: Different wallets handle the home screen balance in various ways. If you setup a default mint, in my opinion it would make sense to show only the balance the user has on that mint because the user has selected a default.

How other wallets handle home screens;
Minibits only shows your balance for a specific mint, and you have to swipe right to view each mint and its unique balance. You can also go to a mint overview screen to see all your balances across different mints.

Cashu.Me and eNuts display your entire collective balance across all mints on the home screen. Cashu.Me shows two balances: a 'grand total' and a 'currently select mint balance'. eNuts doesn't ever show an individual mint balance, even if you assign a default mint. It also shows your total balance across all mints. On eNuts, if you want to see your balances across all mints, you need to go to the mint settings, where a list is displayed.

I think all these solutions could work. Personally, I’m not convinced any one method is better than another. As long as it’s clear what the user is looking at, I believe it comes down to personal preference.

    - file: restore-mint-info-add
      modalImage: restore-mint-info-add@2x
      alt: A mobile wallet interfact that shows a user entering a mint url and the mint name automatically populating.
      caption: Requiring the user to enter the mint url prevents any errors. All other mint information (metadata) can be retrieved from the mint url.
    - file: restore-mint-info
      modalImage: restore-mint-info@2x
      alt: A mobile wallet interfact that shows a list of mints that the user has entered along with the option to add a new mint or continue with the restoration process.
      caption: Provide user with the option to type mint information or scan a mint's QR code.

images_mint-url-edit:
    - file: cashu-mint-details
      modalImage: cashu-mint-details@2x
      alt: A mobile wallet interface showing a mint details screen with an input field to edit the mint url.
      caption: Give the user the option to edit the mint url.
    - file: cashu-mint-url-edit
      modalImage: cashu-mint-url-edit@2x
      alt: A mobile wallet interface showing a user typing a mint url into the mint url input field.

>       caption: Allow the user to manually type the mint url.

Author: Good suggestion. Cashu apps can check if a mint URL is pointing to valid Cashu mint. I'll update this, and I'll include a success and error state validation.

>       caption: Allow the user to manually type the mint url.

Author: Updated from "Update" to "Validate". I'm not sure we need a success and error state screen, if we feel the need to we can address it in a follow up PR.

      caption: Consider the most important information to display to the user about the federation.
    - file: federation-id2
      modalImage: federation-id2@2x
      alt: Federation descriptions display in a wallet interface
      caption: Different federation operator may use the description fields differently. Consider this when designing your federation details screen.
    - file: federation-id3
      modalImage: federation-id3@2x
      alt: Federation contact information display in a wallet interface
      caption: Contact information is optional for the federation operator to include. If there is no contact information, consider not displaying this metadata.
    - file: federation-id4
      modalImage: federation-id4@2x
      alt: Example of important announcements display in a wallet interface
      caption: A message of the day can consists of anything the federation operator wants to tell the user. It can for example be used to announce a new feature or a upcoming maintenance.

images_cashu-id:
    - file: cashu-id1
      modalImage: cashu-id1@2x
      alt: Mint identity and version display in a wallet interface

>       caption: Consider the most important information to display to the user about the mint.

Reviewer 2: What is the "Pubkey" field. Does not seem like something a casual user can make sense of.

Reviewer 2: A version number is also only helpful to highly technical users. Contact info seems more relevant.

Author: Pubkey is not yet used, but it’s in the spec so it needs to be removed. Initially when I included the metadata fields for both Fedimint and Cashu I wanted to provide UI examples showing all the metadata fields the protocol supports. Buit as you pointed out, not all metadata fields are equally important. Some are crucial while some are only relevant to technical users. I will revisit this section and mention this distinction. I’ll focus on UI examples that display the most relevant metadata fields for design, and then include a paragraph about the other fields. 
I'll note that while they may be useful for technical users, they are likely irrelevant for most users. It’s best to carefully consider whether or not to include these fields based on your product’s target audience.

    - file: cashu-id2
      modalImage: cashu-id2@2x
      alt: Mint descriptions display in a wallet interface
      caption: Different mint operator may use the description fields differently. Consider this when designing your mint details screen.
    - file: cashu-id3
      modalImage: cashu-id3@2x
      alt: Mint contact information display in a wallet interface
      caption: Contact information is optional for the mint operator to include. If there is no contact information, consider not displaying this metadata.
    - file: cashu-id4
      modalImage: cashu-id4@2x
      alt: Example of important announcements display in a wallet interface
      caption: A message of the day can consists of anything the mint operator wants to tell the user. It can for example be used to announce a new feature or a upcoming maintenance.
    - file: cashu-id5
      modalImage: cashu-id5@2x

>       alt: Example of supported features display in a wallet interface

Reviewer 2: Seems only useful for highly technical people. Can some of those be actions I can perform? Can I do a "Token state check" if that mint supports that feature?

Author: Yes, token state check is an action users can perform if the mint supports it. It’s quite useful because it allows you to quickly update your token status list without having to re-check each individual token. In eNuts this feature is called ‘check proofs.’ P2PK and multimint payments are also unique actions but no wallet supports multimint payments yet. To your point, listing all of this is probably not useful for most regular users so I’ll remove it.

      caption: Supported features are pulled by indexing a mints information and reading what nuts the mint has implemented.

---
{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/best-practicies.jpg"
   retina = "/assets/images/guide/how-it-works/ecash/best-practicies@2x.jpg"
   mobile = "/assets/images/guide/how-it-works/ecash/best-practicies-mobile.jpg"
   mobileRetina = "/assets/images/guide/how-it-works/ecash/best-practicies-mobile@2x.jpg"
   alt-text = "ecash best practices illustration"
   width = 1600
   height = 700
   layout = "full-width"
%}

<!--

This is a summary page of best design practices for both bitcoin-backed ecash protocols: Cashu and Fedimint.

The design source for screen mock-ups on this page can be found here:
ADD Figma URL

-->

# Ecash design best practices
{: .no_toc }

---

<div class="glossary-toc" markdown="1">
 * Table of contents

> {:toc}

Author: Do you have a guide or example on how to customize the table of contents and link it to specific sections? It seems to automatically display whatever is marked as `##` or H2, but I’d like to create my own shortcuts.

</div>

---

> ## General Ecash Best Practices

Author: I reworked it a bit here. Changed some wording. 
<img width="887" alt="image" src="https://github.com/user-attachments/assets/7737e0aa-b531-45fc-832e-663605c35e21">

> When designing bitcoin-backed ecash applications it's important to prioritize clear and intuitive interfaces that allow users to easily manage their tokens, whether they are minting, sending, or redeeming them.

Author: <img width="836" alt="image" src="https://github.com/user-attachments/assets/a2ee7202-cb20-4889-9c6d-6e0b585ff96f">I want to include something more explicit that mentions the sending practices are the same as those in the existing reference wallet design. Guidelines like entering an amount, displaying fees, and showing a success message still apply. However the key difference is that ecash wallets allow users to generate a payment in the form of an ecash token, which can be shared as a QR code or a string of text. It might be enough to mention this in the text, though I could include an image of a QR code with its text string underneath to show an example. But this image is already shown in the pending tokens section right below.

### Multiple Mint or Federation Display
{% include image-gallery.html pages = page.images_mint-list %}

When displaying multiple mints or federations in a bitcoin-backed ecash application, there are several best practices to consider:

1. **Clear Hierarchy and Organization**: Ensure that mints or federations are organized in a way that is easy for users to navigate. Group mints by relevance, trust level, or frequency of use, and consider using sorting or filtering options to help users find the right mint or federation quickly.

2. **Prominent Metadata Display**: Use metadata fields like mint name, icon, URL, and description, to give users insight into each mint or federations identity. Display this information prominently, so users can easily verify the authenticity and relevance of each mint or federation.

3. **Quick Mint Access**: Provide easy access to mint or federation details directly from the mint or federation list. This allows users to interact with mints or federations more fluidly without needing to navigate through multiple screens.

> #### Pending Tokens

Author: This note on terminology from the transaction page is helpful:
`In most places in the Guide, the term “transaction” and “payment” are used interchangeably. For the remainder of this page, the term transaction refers specifically to on-chain bitcoin transactions, and the term payment refers specifically to payments routed over the lightning network.`

Since ecash is closer to the user experience and use case of lightning, we should use the term ‘payments.’ I went back and updated ‘transactions’ with ‘payments’ where it made sense.

{% include image-gallery.html pages = page.images_pending %}

> It's useful to display generated tokens in the transaction history along with their status—whether pending or redeemed. This helps users quickly assess their transaction status. Providing the ability to expand on a token to view details and take actions like resending or redeeming is also important. This feature becomes particularly valuable if a token was lost or never claimed.

Author: A generated token is simply a token created by the user. Wallets display tokens as 'pending' or 'redeemed' depending on whether they've been redeemed, usually they use an hourglass or a cock icon for 'pending' tokens. Sometimes, a user might issue a token as a gift or payment, but the recipient may not claim it, or they may forget about it. In such cases, if the sender wants to issue a new token from a different mint, it would be useful for the original sender to have the ability to redeem the token they issued and recover the funds.

## Cashu Best Practices

### User Interface and Experience

#### Default Mint
{% include image-gallery.html pages = page.images_mint-default %}

When designing for Cashu, consider allowing users to set a default mint. Since Cashu mints are independently operated, users may prefer to use a trusted mint for their transactions. By allowing them to select a default mint, you simplify their experience and reduce the risk of exposure to a potentially malicious mint. Additionally, consider offering auto-swap preferences.

#### Edit Mint URL & Refresh Mint Settings

{% include image-gallery.html pages = page.images_mint-url-edit %}

It's useful to provide users with the ability to manually update or refresh mint settings in their wallets. While best practices suggest that wallets should auto-refresh and update mint settings periodically, there may be cases where this isn't supported, or a mint undergoes significant updates, such as a URL change. By allowing users to edit the mint URL or refresh settings manually, you ensure that users always have the latest mint configurations.

### Security and Privacy

#### Backup and Restore for Cashu

> The Cashu backup and restoration process is designed to ensure users can securely recover their wallets and maintain access to their bitcoin-backed ecash tokens in the event they switch devices or experience data loss. **Users should only use their recovery seed phrase once.** Repeated use of the seed phrase for restoration can lead to synchronization issues and potential errors. This is due to the fact that each time you restore, you might be dealing with an outdated state of the wallet, which can cause discrepancies in token balances and transactions.

Author: My original understanding was incorrect and was clarified by @thesimplekid. Synchronization issues may only occur if a user is using the same seed phrase across multiple devices. I'll update this with a note that says:

"Running two apps simultaneously with the same seed is not recommended, as it may cause synchronization issues."

{% include tip/close.html %}

#### Backup Process
{% include image-gallery.html pages = page.images_backup %}

1. **Deterministic Wallet with Seed Phrase**: Cashu uses a deterministic wallet model, where all cryptographic keys and bitcoin-backed ecash tokens can be derived from a single seed phrase. This seed phrase can be generated when the wallet is first created, or a at any point in time when the user wants to backup their wallet.

2. **Secure Storage**: Users are advised to store their seed phrase securely. It is crucial to keep this seed phrase in a safe, physical format (such as written on paper) and stored in a secure location to prevent unauthorized access or loss. The best practices in the [Backup & recovery]({{ '/guide/daily-spending-wallet/backup-and-recovery/landing-page/' | relative_url }}) section also apply to Cashu wallets. 

{% include tip/open.html color="blue" icon="info" label="Mint Information Needed" %}

> Keep a record of the mints you are connected to. During the restoration process, you will need to reconnect to these specific mints to re-verify your tokens. Without this information, you will not be able to fully restore your wallet.

Author: Updated to
_"Instruct the user to keep a record of the mints or federations they are connected to. This can be done by writing them down or, since the mint list is not sensitive information, the wallet can offer to save the list of mints on cloud storage. During the restoration process, users will need to reconnect to these specific mints or federations to re-verify their ecash. Without this information, they will not be able to fully restore their wallet."_

{% include tip/close.html %}

#### Restoration Process
{% include image-gallery.html pages = page.images_restore %}

1. **Seed Phrase Entry**: When restoring a wallet, users enter their seed phrase into their wallet. This seed phrase regenerates secrets that may have been used to mint ecash.

2. **Mint(s) Entry**: Users must input the mint URL for each mint they used before. This ensures their wallet can properly reconnect and recreate their tokens.

3. **Verification**: The mint(s) checks these blinded secrets against its records to confirm if it has been seen before and can reissue the signature.

{% include tip/open.html color="blue" icon="info" label="Privacy Considerations During Verification" %}

Verifying tokens after restoration might temporarily compromise their privacy. The mint needs to revalidate the tokens, which can potentially expose the transaction history associated with those tokens. Be mindful of this aspect if privacy is a significant concern for your use case.

{% include tip/close.html %}

### Metadata and Information Management

> #### Cashu Metadata Fields: Explanation and Usage

Author: I get where you're coming from. The only times the screen really changes are for MOTD and supported features, so it could probably be compressed significantly. I'll leave it as is for now, but during the new PR review session, I'll make sure to bring it up and get everyone's thoughts on it.

   alt-text = ""
   width = 250
   height = 541
   layout = "float-right-desktop -background -shadow"
   caption = "Contact information is optional for the mint operator to include. If there is no contact information, consider not displaying this metadata."
%}

- `contact`: An list of contact methods for the mint operator. Consider displaying this in the mint details, using appropriate icons for each contact method (e.g., nostr, email icon, website icon) to make it easy for users to reach out if needed.

</div>

#### Important Announcements

<div class="center" markdown="1">

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/cashu-id4.png"
   retina = "/assets/images/guide/how-it-works/ecash/cashu-id4@2x.png"
   modalImage = "/assets/images/guide/how-it-works/ecash/cashu-id4@2x.png"
   alt-text = "Example of important announcements display in a wallet interface"
   width = 250
   height = 390
   layout = "float-right-desktop -background -shadow"
   caption = "A message of the day can consists of anything the mint operator wants to tell the user. It can for example be used to announce a new feature or a upcoming maintenance."
%}

> - `motd` (Message of the Day): Used for important announcements. Consider displaying this in a prominent, dismissible notification area when users interact with the mint, making sure they don't miss any critical information.

Reviewer 2: There could also be a small notification icon on the home screen.

Author: Yeah, that’s a good point. I think the MOTD should be displayed on the home screen. It’s much less useful if the user has to navigate to a mint details screen to see it, especially since these are timely alerts. I updated the design so the MOTD now appears at the top of the home screen as a dismissible notification.

{% include tip/close.html %}

#### Organizing Cashu Metadata Fields
When designing your Cashu wallet, consider organizing these metadata fields in a logical, user-friendly manner to help users understand and interact with different mints. Here are some suggestions:

- Use the `name` and short `description` in mint listings for quick identification of mints.
- Display a detailed mint information page using `description_long`, `version`, and `contact` details.
- Display the `motd` as a prominent notification when relevant to the specific mint.
- Use the `nuts` information to create a features list that helps users understand each mint's capabilities.

By effectively incorporating these Cashu metadata fields, you can create a more transparent and informative experience for users, helping them make informed decisions about which Cashu mints to trust and use. This is particularly important for Cashu, as users interact directly with individual mints rather than a federation structure like in Fedimint.

### Advance Features

#### Pay to Public Key (P2PK)

{% include image-gallery.html pages = page.images_p2pk %}

> [NUT11](https://github.com/cashubtc/nuts/blob/main/11.md) is a powerful feature that allows bitcoin-backed ecash tokens to be securely locked to another user's public key, which is generated by the recipient's wallet. This ensures that only the intended recipient can redeem the ecash. NUT11 enables secure offline payments, preventing double-spending. Beyond these basics, NUT11 supports advanced use cases like timelocks and multisignature (multisig) setups, where ecash can be conditionally spent or jointly owned by multiple parties. When designing make sure these functionalities are clearly communicated to users, highlighting their practical benefits and flexibility.

Author: I’m not sure I understand. Could you explain it again? Is this potentially a new feature or use case? If so, it might fall outside the scope of the design guide if it doesn’t already exist. As far as I know, no one has included any GPG key functionality in ecash apps.

> [NUT11](https://github.com/cashubtc/nuts/blob/main/11.md) is a powerful feature that allows bitcoin-backed ecash tokens to be securely locked to another user's public key, which is generated by the recipient's wallet. This ensures that only the intended recipient can redeem the ecash. NUT11 enables secure offline payments, preventing double-spending. Beyond these basics, NUT11 supports advanced use cases like timelocks and multisignature (multisig) setups, where ecash can be conditionally spent or jointly owned by multiple parties. When designing make sure these functionalities are clearly communicated to users, highlighting their practical benefits and flexibility.

Reviewer 3: A send-to-contact feature based on Nostr or GPG keys might work as follows: user creates a contact list storing the various user's public keys (or just pick them up from nostr). When sending funds the app UI could offer a send to contact feature what would lock the token to the intended recipient (contact)'s public key so only that they can spend it. I think thats what NUT-11 is for...dunno but I hope this is clearer or just ping me and we can chat.

> [NUT11](https://github.com/cashubtc/nuts/blob/main/11.md) is a powerful feature that allows bitcoin-backed ecash tokens to be securely locked to another user's public key, which is generated by the recipient's wallet. This ensures that only the intended recipient can redeem the ecash. NUT11 enables secure offline payments, preventing double-spending. Beyond these basics, NUT11 supports advanced use cases like timelocks and multisignature (multisig) setups, where ecash can be conditionally spent or jointly owned by multiple parties. When designing make sure these functionalities are clearly communicated to users, highlighting their practical benefits and flexibility.

Author: Sounds good, I'll message you.

#### Auto-Swapping to a Default Cashu Mint
{% include image-gallery.html pages = page.images_auto-swap %}
Wallets can provide the option to automatically swap ecash from an unknown mint to their a default mint. This optional feature can simplify transactions by routing all ecash transactions through the assigned default mint. An auto-swap requires a lightning payment.

{% include tip/open.html color="blue" icon="info" label="Auto-Swapping Involves Fees" %}

It's important to inform users that auto-swapping involves fees and that the amount depends on the amount of ecash they want to swap and the current network conditions.

{% include tip/close.html %}

## Fedimint Design Best Practices

### User Interface and Experience

> #### Federation Status Display

Reviewer 2: Does Cashu not have that feature? Might be good to explain that.

Reviewer: Since cashu is a single mint there is no real status it is either up and working or not, where as a fedimint could have a minatory of guardians down ans still work in a degraded state. 

Cashu does have a message of the day so it could publish a message thats its lightning node is temporarily down so minting and melting is not possible but you can still swap and funds are safe but there isnt a specific spec for this and it would just be a general message so don't think I would include that here

<div class="center" markdown="1">

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/federation-guardian-info"
   retina = "/assets/images/guide/how-it-works/ecash/federation-guardian-info@2x.png"
   modalImage = "/assets/images/guide/how-it-works/ecash/federation-guardian-info@2x.png"
   alt-text = ""
   width = 250
   height = 541
   layout = "float-right-desktop -background -shadow"
   caption = "Example of a guardian status display in a wallet interface."
%}

> Consider using a prominent, color-coded indicator to show the overall status of the federation. This should reflect whether the federation is fully operational, partially degraded, or offline. This helps users quickly assess the health of a federation.

Author: Unfortunately it's not something that Cashu mints support. You can check out this comment by @thesimplekid for more details: https://github.com/BitcoinDesign/Guide/pull/1093#discussion_r1772768036

   caption = "A welcome message being used durring the new user onboarding process."
%}

- `welcome_message`: This field contains a welcome message for new users joining the federation. It could be displayed during the onboarding process to provide users with important information or instructions.

</div>

{% include tip/open.html color="blue" icon="info" label="Flexible Use of Welcome Message" %}
Different federations may choose to display the welcome message in different contexts. Some may use it as a welcome message, others may use it to display the terms of service or other important information.
{% include tip/close.html %}

#### Gateway Information

- `vetted_gateways`: This field contains a list of gateway identifiers that have been vetted by the federation. This information could be used to display trusted gateways to users, helping them make informed decisions about which gateways to use for transactions.

##### What Is a Gateway?
See the [Role of Gateways]({{ '/guide/how-it-works/ecash/fedimint/#role-of-gateways' | relative_url }}) for more information about gateways.

#### Why Showing Gateway Information Might Not Be Necessary:

> 1. **Cognitive Overload**: Most users are not interested in the technical details of how their transactions are processed. Presenting information about gateways, especially vetted versus non-vetted ones, could lead to confusion, as users may not understand what this information means or how to act on it.

Reviewer 2: How about showing something about vetted gateways at the point they become relevant to users? I do think people care, but it doesn't have to be front-loaded or too in-your-face.

Author: I'll have to find out at what point they become relevant to users. I can't think of one. @elsirion any thoughts?

2. **Trust in the System**: In well designed systems, users trust that the service has selected the best pathways for their transactions. By not overwhelming users with technical details, the system can reinforce trust that the backend will manage their transaction securely and efficiently without their intervention.

> 3. **Streamlined Experience**: The goal of good UX is to make the user experience as seamless and intuitive as possible. Introducing gateway information may disrupt this flow by adding unnecessary steps or considerations that the average user doesn't need to think about.

Reviewer 3: Consider if this section could be summarised and replaced by a simple tip as it explains why not to show gateway info.

      caption: Consider the privacy implications of showing the status and name of guardians.
    - file: federation-guardian-ok
      modalImage: federation-guardian-ok@2x
      alt: A mobile wallet interface with a screen that shows a federation experiencing partial connecetivity issues due to one guardian being offline.
      caption: Show the user the current status of the federation.
    - file: federation-guardian-bad
      modalImage: federation-guardian-bad@2x
      alt: A mobile wallet interface with a screen that shows an unreliable federation and their health status in the form of a red progress bar along with a message that the federation is offline.
      caption: Clearly inform a user if a federation is offline.

images_guardian-expiry:
    - file: federation-guardian-expiry
      modalImage: federation-guardian-expiry@2x
      alt: A mobile wallet interface with a screen that shows a countdown timer and a message that the federation will shut down in 15 days.

>       caption: Federation expiration timestamp displayed as a countdown alter when less 30 days remaining.

Reviewer 4: Nit: "alter" -> "alert"

Also it would be useful for federations to be able to extend their expiration time (i.e. they're promising to not shut down before a certain date, but not promising that they will _definitely_ shut down at that date. They might tack on more time later)

It's important to inform users that auto-swapping involves fees and that the amount depends on the amount of ecash they want to swap and the current network conditions.

{% include tip/close.html %}

## Fedimint design guidelines

### User interface and experience

#### Federation status display

<div class="center" markdown="1">

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/federation-guardian-info"
   retina = "/assets/images/guide/how-it-works/ecash/federation-guardian-info@2x.png"
   modalImage = "/assets/images/guide/how-it-works/ecash/federation-guardian-info@2x.png"
   alt-text = ""
   width = 375
   height = 812
   layout = "float-right-desktop -background -shadow"
   caption = "Example of a guardian status display in a wallet interface."
%}

Consider using a prominent, color-coded indicator to show the overall status of the federation. This should reflect whether the federation is fully operational, partially degraded, or offline. This helps users quickly assess the health of a federation.

</div>
#### Guardian Status Indicators
{% include image-gallery.html pages = page.images_guardian-status %}

> Consider displaying real-time indicators of each guardian's status, such as the connection status and last activity. This information builds trust by keeping users informed about the reliability and performance of the guardians managing their funds.

Author: Updated so "last active" is only visible for offline guardians.

### Security and privacy

#### Privacy considerations in guardian information display
Fedimint automatically generates random names for guardians during setup. While users can change these names, it’s not recommended. When displaying guardian information, consider using code names like "Guardian 1" and randomly generated identicons for each guardian. This helps protect user and federation privacy, particularly for invite-only mints. Guardian information should only be visible to users who have joined the federation.

### Metadata and information management

#### Federation metadata fields: Explanation and usage

> Federations can provide configuration and metadata to users. These metadata fields are consensus-relevant, meaning they must be consistent across all federation members. This ensures that users can rely on their accuracy. Let's explore the core metadata fields defined in the Fedimint protocol:

Reviewer 4: Can we link to the fedimint documentation here?

https://github.com/fedimint/fedimint/blob/master/docs/meta_fields/README.md
